src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -342,6 +342,11 @@ where
         &mut self,
         commands: Vec<(GlobalId, Vec<Update<Self::Timestamp>>, Self::Timestamp)>,
     ) -> Result<(), StorageError> {
+        for (_, updates, upper) in commands.iter() {
+            for update in updates {
+                assert!(update.timestamp.less_than(upper), "invalid append batch");
+            }
+        }
```

Comment:
Could you return a `StorageError` instead of panicking?

Reply:
Yep, fixed